src/gss_names.c

```diff
@@ -281,6 +281,84 @@ uint32_t gssntlm_import_name(uint32_t *minor_status,
                                        output_name);
 }
 
+size_t gssntlm_get_attrs_count(const struct gssntlm_name_attribute *attrs)
+{
+    size_t attrs_count = 0;
+    if (!attrs) {
+        return 0;
+    }
+    while (attrs[attrs_count].attr_name != NULL) {
+        attrs_count++;
+    }
+
+    return attrs_count;
```

Comment:
Again, you sacrifice efficiency - no need to check 'attrs != NULL' per each iteration of the loop.

Reply:
the compiler will handle this, example of assembly generated by gcc -O2 for this code:
`B+>6               for (c = 0; argv != NULL && argv[c] != 0; c++);`
```
B+>0x401020 <main>                                 xor    eax,eax                
      0x401022 <main+2>                               test   rsi,rsi
      0x401025 <main+5>                               jne    0x401034 <main+20>     
      0x401027 <main+7>                               jmp    0x40103b <main+27>
      0x401029 <main+9>                               nop    DWORD PTR [rax+0x0]    
      0x401030 <main+16>                              add    rax,0x1
      0x401034 <main+20>                              cmp    BYTE PTR [rsi+rax*1],0x0
      0x401038 <main+24>                              jne    0x401030 <main+16> 
```

As you can see the first line sets c to 0, then there is a single test for argv not to be null (the jne instruction)[1][2], and then the loop is between main+16 - main+24 which is the three instruction you would expect, no additional testing of argv because the compiler knows there is no instruction changing it.

So do not worry about efficiency, please use the compact form, it is idiomatic in or code. 
[1] note that because the compiler know the branch will be predicted it uses a jne+jmp combination instead of a simple je, this way the CPU predictor will make faster code even though it looks a je could be faster
[2] not that at -O3 it uses two je instructions instead and two cmp, and also maintains the nop, it was just less readable so I chose to show -O2, in both cases the test was done only once before the loop.

Reply:
Wow, you have done significant research to proof your concept... Ok, will fix it as you want

Reply:
took no more than 10 min to write code disassemble and paste in github comment [5 minutes where spent fighting the comment formatting :-D ]